market-contract/src/main.sw

```diff
@@ -971,103 +906,176 @@ fn match_order_internal(
+        // Handle partial or full order fulfillment and rewards
+        update_order_storage_and_reward(
+            amount,
+            s_order,
+            s_id,
+            amount,
+            b_order,
+            b_id,
+            &mut matcher_reward,
+        )
+        // If orders do not match by type or asset, no match occurs
     } else {
         (MatchResult::ZeroMatch, ZERO_B256, matcher_reward)
     }
 }
 
+#[storage(read, write)]
+fn update_protocol_fee(amount: u64) {
+    storage
+        .total_protocol_fee
+        .write(storage.total_protocol_fee.read() + amount);
+}
+
+#[storage(read, write)]
+fn update_order_storage_and_reward(
+    amount0: u64,
+    ref mut order0: Order,
+    id0: b256,
+    amount1: u64,
+    ref mut order1: Order,
+    id1: b256,
+    matcher_reward: &mut u64,
+) -> (MatchResult, b256, u64) {
+    // Case where the first order is completely filled
+    if amount0 == order0.amount {
+        update_protocol_fee(order0.protocol_fee);
+
+        *matcher_reward += order0.matcher_fee.as_u64();
+        remove_order(order0.owner, id0);
+    }
+    // Case where the second order is completely filled
+    if amount1 == order1.amount {
+        update_protocol_fee(order1.protocol_fee);
+
+        *matcher_reward += order1.matcher_fee.as_u64();
+        remove_order(order1.owner, id1);
+    }
+    // Case where the first order is partially filled
+    if amount0 != order0.amount {
+        let fee = order0.protocol_fee * amount0 / order0.amount;
+        update_protocol_fee(fee);
+        order0.protocol_fee -= fee;
+
+        let order_matcher_reward = order0.matcher_fee.as_u64() * amount0 / order0.amount;
+        *matcher_reward += order_matcher_reward;
+        order0.matcher_fee -= order_matcher_reward.try_as_u32().unwrap();
+        order0.amount -= amount0;
+        storage.orders.insert(id0, order0);
+        return (MatchResult::PartialMatch, id0, *matcher_reward);
+        // Case where the second order is partially filled
+    } else if amount1 != order1.amount {
+        let fee = order1.protocol_fee * amount1 / order1.amount;
+        update_protocol_fee(fee);
+        order1.protocol_fee -= fee;
+
+        let order_matcher_reward = order1.matcher_fee.as_u64() * amount1 / order1.amount;
+        *matcher_reward += order_matcher_reward;
+        order1.matcher_fee -= order_matcher_reward.try_as_u32().unwrap();
+        order1.amount -= amount1;
+        storage.orders.insert(id1, order1);
+        return (MatchResult::PartialMatch, id1, *matcher_reward);
+    }
+    // Case where both orders are fully matched
+    (MatchResult::FullMatch, ZERO_B256, *matcher_reward)
+}
+
+#[storage(read, write)]
+fn emit_match_events(
+    id0: b256,
+    order0: Order,
+    amount0: u64,
+    id1: b256,
+    order1: Order,
+    amount1: u64,
+    matcher: Identity,
+    match_price: u64,
+) {
+    // Emit events for the first order
+    log_order_change_info(
```

Comment:
Ask Alex should we remove log_order_change_info. IT was created for indexer debug 3 months ago

Reply:
Will do.